### ISSUE TITLE
Ideas for basic features

### ISSUE THREAD

**Issue by vlazar** [vlazar/a-priori-tizer#1]
Huge list of unread notifications for GitHub PRs became to big and scary to deal with one day. Small PRs are easier to review, so there is a need to find small PRs among others and deal with them first.

Problem:
- notifications do not show PR size and even if they do, one can't sort emails by PR size
- there is no way to sort PR's by number of changes on GitHub currently

Solution:

Show a list of PRs sorted by PR size in OS X menubar. No new notifications, no other noise, just a list of prioritized items to deal with one by one.

Some ideas:
- PRs from multiple repos in one list
- one line for each PR, at least for now
- show only open PRs obviously
- sort PRs by number of changes (additions + deletions)
- group PRs (small/medium/large PRs separated by horizontal line)
- show PR title
- on click open GitHub PR page
- use different color for PRs from different groups
- show number of changes in PR (additions + deletions)
- show some PR status icon (e.g. mergeable/non mergeable)
- show build status icon (success/failure)
- in OS X menubar show icon with number of PRs (e.g. number of PRs in less scary group)

Example output with PR groups:

PRs in each group are shown in green, orange and red color correspondingly. OS X menubar shows ● 2 in green as there are 2 small PRs. The ● for PR means it's mergeable, while ○ means PR is not mergeable.

```
● 2
---------------------------------------------------------------------------
●  Allow container to be placed on a different object
●  Up version of bower.json and package.json to 0.1.0 in order to npm an…
---------------------------------------------------------------------------
●  Fix #16808. Input loses focus if item selected with mouse.
○  Ensure highlighting matches the filter method chosen
●  Add toggle API method
○  Better accessibility
○  Adding keepListItems configuration property.
○  Implement additional event for click on autocompletion list - fix #16712
○  Allow user to specify pre-existing dropdown element via options param
○  Allow preexisting wrapper and custom class names
●  Feature/handle array of objects
---------------------------------------------------------------------------
○  Keep original list order
●  add key value support
---------------------------------------------------------------------------
```

**Comment by vlazar** [comment]
This is a list of info available about each GitHub PR https://developer.github.com/v3/pulls/#get-a-single-pull-request

**Comment by LeaVerou** [comment]
In general, try to avoid icons if meaning is not clear immediately. They add to the amount of knowledge someone needs before using the UI.
In this case, there is an obvious color coding when PR is not mergeable: Gray. You could even use the icon Github shows for non mergeable PRs.
Regarding size, I would like the small icon for additions/deletions that Github shows. External consistency (in this case with Github) is important for a good UI, so instead of inventing your own conventions, try to use the ones people are already familiar with from Github.
No need to show something if PR is mergeable. Just assume PRs with no indication are mergeable and only show an indication if they are not. Don't bloat the UI for redundant information.
It's also useful to know if maintainers have replied to PRs and perhaps if there were other comments after the last maintainer comment. Since Github doesn't have an existing convention for that, this is where you have to invent your own. Perhaps different colored speech bubbles with number of comments next to them or inside them? Red if no maintainer has replied, orange if there are comments after the last maintainer comment etc.
Does this help?

**Comment by vlazar** [comment]
Thanks for feedback Lea, it sure helps!
So, the list of features:
- color code PR merge-able status (gray might confuse though, as non clickable items in OS X menu are gray)
- show additions/deletions like GitHub, instead of number of changes
- show if there are replies from maintainer to this PR
- show if there are comments since after the last maintainer comment
- show number of comments in speech bubble (red - maintainer has replied, orange - there are comments after the last maintainer comment, etc.)

I've started implementing this as a [BitBar](https://github.com/matryer/bitbar) plugin, so I can't show any icon I want currently, only Unicode/Emoji. There is a feature request for icons support for BitBar, so this might change in the future.

I have have chosen BitBar since it's lightweight and looks like it will be even more popular soon. Another option would be to use Electron, however it's not that lightweight. Going native is to much time for me, I'm not an Objective C/Swift guy.

**Comment by vlazar** [comment]
> Regarding size, I would like the small icon for additions/deletions that Github shows.

Did you mean showing 5 green/red squares, just numbers like +10 -11, or both squares & numbers?

**Comment by LeaVerou** [comment]
Actually, now that I looked more closely at the squares, I don't think it's that useful.
I think, as a maintainer, I would primarily need to know number of additions/deletions (perhaps even as a sum).

**Comment by vlazar** [comment]
Wouldn't it be nice, if we show for each PR how much time it would require and order PRs by that time? Like on some blogs there is an indicator saying it's a 5min read and 3 minutes left. Something like that.

I wonder if it's easy/possible to come up with reasonable metric for PRs based on additions/deletions, number of comments, author of PR and if this author has merged PRs already and any other info we can get via GitHub API and calculate based on everything we know.  Preferably this should be customizable, as each person would most likely have a different vision of what is important and time for PR for each person can be calculated differently based on different criteria.

**Comment by LeaVerou** [comment]
Yes, it would be great, but the only metric I can think of is number of additions & deletions...

**Comment by LeaVerou** [comment]
Yes, I really like this idea! But it should have a good default, as most people won't customize it (that's true for every customizable setting)

**Comment by vlazar** [comment]
I'll think on this more as well. Having some heuristic algorithm adopting to user automatically would be a dream :)

Currently I plan to use Ruby for implementation, since it's installed by default on any OS X, and Node.js requires installation. Node.js also requires node path in bash script, which can be different. I've also tried to use JXA for OS X but it looks like it has issues with BitBar, so for now I'll use Ruby.
